main.go

```diff
@@ -162,47 +133,105 @@ func main() {
 			return err
 		}
 		return nil
-	}); err != nil {
-		log.Errorf("Nuget restore failed, error: %s", err)
-		os.Exit(1)
+	})
+}
+
+// collectCaches collects the caches based on the allowed config.
+// For more information about caches please read: https://docs.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders
+func collectCaches(allowCaching string) error {
+	nuGetCache := cache.New()
+	switch allowCaching {
+	case cacheInputNone:
+		return nil
+	case cacheInputlocal:
+		collectLocalCaches(&nuGetCache)
+	case cacheInputGlobal:
+		collectGlobalCaches(&nuGetCache)
 	}
+	return nuGetCache.Commit()
+}
 
-	// Collecting caches
-	fmt.Println()
-	log.Infof("Collecting NuGet cache...")
+// collectHTTPCaches collects the HTTP cache.
+func collectHTTPCaches(nuGetCache *cache.Cache) {
```

Comment:
return the http cache path instead of modifying a function parameter, pls avoid using functions with a side effect, it is hard to follow how the data flows.

Reply:
Updated code. However, please note:
The code looked like this before the update
```
func collectCaches(allowCaching string) error {
 	nuGetCache := cache.New()
 	switch allowCaching {
 	case cacheInputNone:
 		return nil
 	case cacheInputlocal:
 		collectLocalCaches(&nuGetCache)
 	case cacheInputGlobal:
 		collectGlobalCaches(&nuGetCache)
 	}		 	}
 	return nuGetCache.Commit()
 }
```

Now it looks like this:
```
func collectCaches(cacheLevel string, basePth string) error {
	nuGetCache := cache.New()
	switch cacheLevel {
	case cacheInputNone:
		return nil
	case cacheInputlocal:
		localCaches, err := collectLocalCaches(basePth)
		if err != nil {
			log.Warnf("Error occurred while getting local cache. Error: %s", err)
		}
		for _, lcItem := range localCaches {
			nuGetCache.IncludePath(lcItem)
		}
	case cacheInputGlobal:
		nuGetCache.IncludePath(collectGlobalCaches())
	case cacheInputAll:
		localCaches, err := collectLocalCaches(basePth)
		if err != nil {
			log.Warnf("Error occurred while getting all cache. Error: %s", err)
		}
		for _, lcItem := range localCaches {
			nuGetCache.IncludePath(lcItem)
		}
		nuGetCache.IncludePath(collectGlobalCaches())
	}
	return nuGetCache.Commit()
}
```

It became much harder to read. I agree, that it is harder to follow the flow when you are modifying a function param, but not much and I think this won't be a problem, if you choose a very understandable method name and also add documentation. Also, according to good coding principles, methods should be very short, ideally just a few lines.

Reply:
Yeah it was easiert to read, but in general functions with side effects are problematic, not just the data flow, but also testing is much harder. 

I think you can both avoid modifying function parameters, and keep the code readble. What about this:

```
func collectCaches(cacheLevel string, basePth string) error {
	if cacheLevel == cacheInputNone {
		return nil
	}

	nuGetCache := cache.New()

	if cacheLevel == cacheInputlocal || cacheLevel == cacheInputAll {
		localCaches, err := collectLocalCaches(basePth)
		if err != nil {
			log.Warnf("Error occurred while getting local cache: %s", err)
		}
		for _, lcItem := range localCaches {
			nuGetCache.IncludePath(lcItem)
		}
	}

	if cacheLevel == cacheInputGlobal || cacheLevel == cacheInputAll {
		nuGetCache.IncludePath(collectGlobalCaches())
	}

	return nuGetCache.Commit()
}
```

Reply:
As these methods (collectLocalCaches, collectGlobalCaches) work with files, ideally they should be mocked for unit tests, and we should test the collectCaches method. For this reason I updated that code, but passing the cache as arg would not make testing harder in this case.

I decided to go with the switch statement, it is longer than this new form, but it is much more structured.